File: prac-sem6/daa_adbms_codes/daa cpp/fractional_knapsack.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

// Structure to represent an item with weight and value
struct Item {
    int weight;
    int value;
};

// Comparator function to sort items by value/weight ratio in descending order
bool compare(Item a, Item b) {
    double r1 = (double)a.value / a.weight;
    double r2 = (double)b.value / b.weight;
    return r1 > r2;
}

// Function to calculate the maximum value for the fractional knapsack problem
double fractionalKnapsack(int W, vector<Item>& items) {
    // Sort items by value-to-weight ratio
    sort(items.begin(), items.end(), compare);

    double maxValue = 0.0;  // Maximum value that can be attained
    int currentWeight = 0;  // Current weight in the knapsack

    for (auto item : items) {
        // If adding the entire item won't exceed capacity
        if (currentWeight + item.weight <= W) {
            currentWeight += item.weight;
            maxValue += item.value;
        }
        // If we can only take a fraction of the remaining item
        else {
            int remain = W - currentWeight;
            maxValue += item.value * ((double)remain / item.weight);
            break;
        }
    }

    return maxValue;
}

int main() {
    int n, W;
    cout << "Enter the number of items: ";
    cin >> n;
    cout << "Enter the capacity of the knapsack: ";
    cin >> W;

    vector<Item> items(n);
    cout << "Enter weight and value for each item:" << endl;
    for (int i = 0; i < n; i++) {
        cin >> items[i].weight >> items[i].value;
    }

    double maxProfit = fractionalKnapsack(W, items);
    cout << "Maximum value in Knapsack = " << maxProfit << endl;

    return 0;
}

File: prac-sem6/daa_adbms_codes/daa codes/Knapsack/FKnapsack/FractionalKnapsack.java
import java.util.Arrays;
import java.util.Comparator;
import java.util.Scanner;

public class FractionalKnapsack {

    static class Item {
        int profit, weight;

        Item(int x, int y) {
            this.profit = x;
            this.weight = y;
        }
    }

    static class ComparatorByRatio implements Comparator<Item> {
        @Override
        public int compare(Item a, Item b) {
            double ratio1 = (double) a.profit / a.weight;
            double ratio2 = (double) b.profit / b.weight;
            if (ratio1 < ratio2) {
                return 1;  // Return 1 if ratio1 is smaller (for descending order)
            } else if (ratio1 > ratio2) {
                return -1; // Return -1 if ratio1 is greater (for descending order)
            } else {
                return 0;
            }
        }
    }

    static double fractionalKnapsack(int w, Item[] arr, int n) {
        Arrays.sort(arr, new ComparatorByRatio()); // Sort items by profit/weight ratio

        int currentWeight = 0;
        double finalProfit = 0.0;

        for (int i = 0; i < n; i++) {
            if (currentWeight + arr[i].weight <= w) {
                // Take the full item
                currentWeight += arr[i].weight;
                finalProfit += arr[i].profit;
                System.out.println("Taking full item " + (i+1) + " with profit: " + arr[i].profit + " and weight: " + arr[i].weight);
            } else {
                // Take the fraction of the remaining item
                int remainingCapacity = w - currentWeight;
                finalProfit += arr[i].profit * ((double) remainingCapacity / arr[i].weight);
                System.out.println("Taking fractional part of item " + (i+1) + " with profit: " + (arr[i].profit * ((double) remainingCapacity / arr[i].weight))) ;
                break;
            }
        }

        return finalProfit;
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        System.out.println("Enter number of items:");
        int n = sc.nextInt();
        Item[] arr = new Item[n];

        // Input the profit and weight of items
        for (int i = 0; i < n; i++) {
            System.out.println("Enter profit and weight for item " + (i + 1) + ": ");
            int profit = sc.nextInt();
            int weight = sc.nextInt();
            arr[i] = new Item(profit, weight);
        }

        System.out.println("Enter weight capacity: ");
        int w = sc.nextInt();

        FractionalKnapsack obj = new FractionalKnapsack();
        double ans = obj.fractionalKnapsack(w, arr, n);
        System.out.println("Maximum profit is: " + ans);

        sc.close();
    }
}
